### PR TITLE
Use Parametrized test for Number Search

### LINKED ISSUE #545: Use parametrized tests for Number Search
or use proper parameterized tests? may be an overkill though.

_Originally posted by @jingtang10 in https://github.com/google/android-fhir/pull/436#discussion_r650080850_

### PR DESCRIPTION
**Pull request by Anshul-9923**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #545 

**Description**
I have created one file NumberSearch.kt in which I have created a parameterized test for function search_filter_number_lessThanEquals. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [ ] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [ ] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [ ] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by Anshul-9923** [comment]
@jingtang10 Please review my code if any changes are required let me know. For now, I have created a parameterized test for only 1 function after reviewing this I will create for other functions as well.

File: engine/src/test/java/com/google/android/fhir/search/SearchTest.kt [hunk 1]
```diff
@@ -124,8 +124,8 @@ class SearchTest {
 
   @Test
   fun search_filter_number_equals() {
-    /* x contains pairs of values and their corresponding range (see BigDecimal.getRange() in
-    MoreSearch.KT) */
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/SearchTest.kt:128]: change not required

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  private val expected2: String
+) {
+    companion object {
+    @JvmStatic
+    @Parameterized.Parameters
+    fun data() : Collection<Array<Any>> {
+      return listOf(
+        arrayOf(BigDecimal("100.00"),""" 
+        SELECT a.serializedResource
+        FROM ResourceEntity a
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:27]: Can this string not be a part of the test? You're repeating the same string in all the cases.

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+          ResourceType.RiskAssessment.name,
+          ResourceType.RiskAssessment.name,
+          RiskAssessment.PROBABILITY.paramName,
+          BigDecimal("1e2").toDouble()
+        ))
+      )
+    }
+  }
+  @Test
+  fun search_filter_number_lessThanEquals() {
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:85]: please add the other cases as well.

**Review by epicadk — CHANGES_REQUESTED** [review]

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 3 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:85]: @epicadk other cases mean other functions search_filter_number_notEquals, search_filter_number_greaterThanEquals,etc.

File: engine/src/test/java/com/google/android/fhir/search/SearchTest.kt (hunk 1 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/SearchTest.kt:128]: I will undo the change in the next commit.

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 2 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt]: Yes, I did that by creating a variable.

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+@RunWith(Parameterized::class)
+class NumberSearch(
+  private val num: BigDecimal,
+  private val expected1: String,
+  private val expected2: String
+) {
+    companion object {
+    @JvmStatic
+    @Parameterized.Parameters
+    fun data() : Collection<Array<Any>> {
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:23]: The only parametrised quantities need to be the number itself and its upper and lower bounds. The rest of these don't change and can be constants in the test.

**Review by epicadk — CHANGES_REQUESTED** [review]

**Comment by Anshul-9923** [comment]
@epicadk I have committed the changes if now the parameterized test for search_filter_number_lessThanEquals is all right then I would create a parameterized test for other functions as well.

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+@RunWith(Parameterized::class)
+class NumberSearch(
+  private val num: BigDecimal
+) {
+    companion object {
+    @JvmStatic
+    @Parameterized.Parameters
+    fun data() : Collection<Array<Any>> {
+      return listOf(
+        arrayOf(BigDecimal("100.00")),
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:23]:
```suggestion
        arrayOf(BigDecimal("100.00"),BigDecimal("99.995"),BigDecimal("100.005"))),
```

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        AND a.resourceId IN (
+        SELECT resourceId FROM NumberIndexEntity
+        WHERE resourceType = ? AND index_name = ? AND index_value <= ?
+        )
+        """.trimIndent())
+    assertThat(query.args).isEqualTo(listOf(
+      ResourceType.RiskAssessment.name,
+      ResourceType.RiskAssessment.name,
+      RiskAssessment.PROBABILITY.paramName,
+      BigDecimal("100.00").toDouble()
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:57]: Use the parameter here as well.

**Review by epicadk — CHANGES_REQUESTED** [review]
Thanks for this, better now. Please add the remaining tests as well after making these changes.
refer https://www.hl7.org/fhir/search.html#number

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 5 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:23]: @epicadk Like this should I change other list elements also?

**Comment by Anshul-9923** [comment]
@epicadk I have added the remaining tests and done changes as told by you. PTAL!

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 6 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt]: Done

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+        """.trimIndent()
+      )
+
+    assertThat(query.args)
+      .isEqualTo(
+        listOf(
+          ResourceType.RiskAssessment.name,
+          ResourceType.RiskAssessment.name,
+          RiskAssessment.PROBABILITY.paramName,
+          num.toDouble()
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:364]: Have you tried running the tests locally. This doesn't look like it's going to work. The number of expected args and the args do not match.

Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:364]: Yes Sir I tried but it shows tests events were not received. Can you please tell me which tests will not be going to work?

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import com.google.common.truth.Truth.assertThat
+import java.lang.IllegalArgumentException
+import org.junit.Assert
+
+
+@RunWith(Parameterized::class)
+class NumberSearch(
+  private val num: BigDecimal
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:18]: the test events are received probably because the class is expecting 1 argument but you are passing 3. Please have 3 args here something like , number, higherBound, lowerBound

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 7 above)
Comment (epicadk) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:364]: > Yes Sir I tried but it shows tests events were not received. Can you please tell me which tests will not be going to work?

No need for the formalities. 😅

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 8 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt]: @epicadk Resolved the problem and did the required changes. PTAL

**Comment by epicadk** [comment]
Looks good, but I still think a few of the tests will fail. Please look into those.

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 9]
```diff
@@ -0,0 +1,414 @@
+package com.google.android.fhir.search
+
```
Comment (epicadk) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:2]: Please run spotless using `gradlew sA`

**Review by epicadk — DISMISSED** [review]

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 9 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:2]: @epicadk Should I have to set up anything else before running this command because I tried to run this command it shows the build failed in the 
Screenshot (478)
android studio terminal.

Comment (epicadk) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:2]: https://intellij-support.jetbrains.com/hc/en-us/community/posts/360008464899-What-s-the-smart-commands-execution

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt [hunk 10]
```diff
@@ -0,0 +1,414 @@
+package com.google.android.fhir.search
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt:1]: Test file name should end with "Test".

Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt]: Done

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 9 above)
Comment (Anshul-9923) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt]: @epicadk I tried to understand but didn't able to get what you want to say to resolve this issue.

**Review by jingtang10 — COMMENTED** [review]
are you going to remove the tests in the `SearchTest.kt` file? @epicadk what do you think? should we have parameterized tests for all types?

**Comment by epicadk** [comment]
> are you going to remove the tests in the `SearchTest.kt` file? @epicadk what do you think? should we have parameterized tests for all types?

Yup I think we would cover a lot of edge cases that way.

File: engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt (hunk 9 above)
Comment (epicadk) [reply engine/src/test/java/com/google/android/fhir/search/NumberSearch.kt]: press ctrl+enter after you type `gradlew sA`

**Comment by jingtang10** [comment]
@epicadk @omarismail94 can you two please take another look at this PR. any more changes required or do we want to merge this?

@Anshul-9923 

thanks

**Comment by omarismail94** [comment]
Hey @Anshul-9923 -  are you still interested on working on this PR?

I pulled your change on my machine and ran :

```
./gradlew :engine:testDebugUnitTest \
--tests "com.google.android.fhir.search.NumberSearchTest"
```

to test the changes you made and they failed. Could you take a look and fix them?

You also need to format the code before uploading your new commit as it is failing right now. See the CI/ build logs here:
https://github.com/google/android-fhir/runs/6939769862?check_suite_focus=true

To format the code, run:

```
./gradlew spotlessApply
```

and you will see that it will format the files you are changing for you

**Comment by omarismail94** [comment]
@epicadk @jingtang10 

can you review the changes I made? All errors now resolved when I run `./gradlew check` locally. I also removed the non-parameterised version of the tests in [SearchTest.kt](https://github.com/google/android-fhir/pull/1153/files#diff-f1c6d500d35226a0690a062de7822f3136aae2bf6de227b120ada65a818ffe89)

File: engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+          filter(
+            RiskAssessment.PROBABILITY,
+            {
+              prefix = ParamPrefixEnum.GREATERTHAN_OR_EQUALS
+              value = num
+            }
+          )
+        }
+        .getQuery()
+    assertThat(query.query)
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt:178]: `query.query` looks strange - so maybe the variable should be renamed to search. so you get `search.query`.

same for the rest of the test

File: engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+import org.junit.Assert
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.ParameterizedRobolectricTestRunner
+import org.robolectric.annotation.Config
+
+/** Unit tests for when [NumberParamFilterCriterion] used in [MoreSearch]. */
+@RunWith(ParameterizedRobolectricTestRunner::class)
+@Config(sdk = [Build.VERSION_CODES.P])
+class NumberSearchTest(
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt:35]:
```suggestion
class NumberSearchParameterizedTest(
```

File: engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+    """
+    SELECT a.serializedResource
+    FROM ResourceEntity a
+    WHERE a.resourceType = ?
+    AND a.resourceUuid IN (
+    SELECT resourceUuid FROM NumberIndexEntity
+    """.trimIndent()
+
+  @Test
+  fun search_filter_number_equals() {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt:62]: kinda feel like now that these are in the same file, we don't need to say "search" and "filter" and "number" in the test name any more. what about using this more readable style:

`should search equal values`

https://kotlinlang.org/docs/coding-conventions.html#names-for-test-methods

File: engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+          ResourceType.RiskAssessment.name,
+          RiskAssessment.PROBABILITY.paramName,
+          lowerBound.toDouble(),
+          upperBound.toDouble()
+        )
+      )
+  }
+
+  @Test
+  fun search_filter_number_greater() {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt:132]: `should search values greater than a number`

etc

File: engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+    SELECT a.serializedResource
+    FROM ResourceEntity a
+    WHERE a.resourceType = ?
+    AND a.resourceUuid IN (
+    SELECT resourceUuid FROM NumberIndexEntity
+    """.trimIndent()
+
+  @Test
+  fun search_filter_number_equals() {
+    /* x contains pairs of values and their corresponding range (see BigDecimal.getRange() in
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/search/NumberSearchTest.kt:63]: does this comment still name sense? can you rewrite?

**Review by jingtang10 — APPROVED** [review]
thanks omar! some nits but overall looking good.

**Comment by jingtang10** [comment]
disabled auto merge - please address the comments before merging.

**Comment by omarismail94** [comment]
@jingtang10 thanks for the review! Addressed comments and will wait for tests to pass before merging

**Review by omarismail94 — APPROVED** [review]
